pkg/updater/updater.go

```diff
@@ -131,97 +117,185 @@ func lockGroup(ctx context.Context, client gcs.ConditionalClient, path gcs.Path,
+	r, attrs, err := client.Open(ctx, configPath)
+	if err != nil {
+		if !isPreconditionFailed(err) {
+			err = fmt.Errorf("read: %v", err)
+		}
+		return 0, nil, err
 	}
-	// run the garbage collector after each group to minimize
-	// extraneous memory usage.
-	runtime.GC()
-	return nil
+	cfg, err := config.Unmarshal(r)
+	if err != nil {
+		return 0, nil, fmt.Errorf("unmarshal: %v", err)
+	}
+	var configGen int64
+	if attrs != nil {
+		configGen = attrs.Generation
+	}
+
+	var groups []*configpb.TestGroup
+	if group != "" { // Just a specific group
+		tg := config.FindTestGroup(group, cfg)
+		if tg == nil {
+			return 0, nil, errors.New("group not found")
+		}
+		groups = []*configpb.TestGroup{tg}
+	} else { // All groups
+		groups = cfg.TestGroups
+	}
+
+	generations := make(map[string]int64, len(groups))
+
+	q.Init(groups, time.Now())
+
+	if len(groups) > 0 {
+		paths, err := gridPaths(configPath, gridPrefix, groups)
+		if err != nil {
+			return configGen, nil, err
+		}
+		attrs := gcs.Stat(ctx, client, 20, paths...)
+		updates := make(map[string]time.Time, len(attrs))
+		now := time.Now()
+		for i, attrs := range attrs {
+			name := groups[i].Name
+			switch {
+			case attrs.Attrs != nil:
+				updates[name] = attrs.Attrs.Updated.Add(freq)
+				generations[name] = attrs.Attrs.Generation
+			case attrs.Err == storage.ErrObjectNotExist:
+				updates[name] = now
+				generations[name] = 0
+			default:
+				// no change
+			}
+		}
+		q.FixAll(updates)
+	}
+	return configGen, generations, nil
 }
 
-// Update performs a single update pass of all all test groups specified by the config.
-func Update(parent context.Context, client gcs.ConditionalClient, mets *Metrics, configPath gcs.Path, gridPrefix string, groupConcurrency int, group string, updateGroup GroupUpdater, write bool) error {
-	defer growMaxUpdateArea()
+// Updates test groups with the specified freq.
+//
+// Filters down to a single group when set.
+// Returns after all groups updated once if freq is zero.
+func Update(parent context.Context, client gcs.ConditionalClient, mets *Metrics, configPath gcs.Path, gridPrefix string, groupConcurrency int, group string, updateGroup GroupUpdater, write bool, freq time.Duration) error {
 	ctx, cancel := context.WithCancel(parent)
 	defer cancel()
 	log := logrus.WithField("config", configPath)
-	cfg, err := config.ReadGCS(ctx, client, configPath)
+
+	var q config.TestGroupQueue
+
+	gen, generations, err := updateTestGroups(ctx, client, &q, configPath, gridPrefix, group, freq)
 	if err != nil {
-		return fmt.Errorf("read config: %w", err)
+		return err
 	}
-	log.WithField("groups", len(cfg.TestGroups)).Info("Updating test groups")
-
-	groups := make(chan configpb.TestGroup)
+	var lock sync.Mutex
 	var wg sync.WaitGroup
+	wg.Add(groupConcurrency)
 	defer wg.Wait()
-	defer close(groups)
-
-	var generations map[string]int64
-
+	channel := make(chan *configpb.TestGroup) // TODO(fejta): pass into this function
```

Comment:
`// TODO(fejta): pass into this function` Why do we want this?

Reply:
Done